### PR TITLE
types(core): add type hints in core for better inference

### PR DESCRIPTION
**Pull request by oscarbenjamin**
#### References to other Issues or PRs

Follows gh-27217 (merged) and gh-27218 (outstanding).

#### Brief description of what is fixed or changed

Add more type hints so that types can be inferred for common operations with the most commonly used expression types. In particular add overloads for subs so that it is possible to distinguish between Expr and Basic. Also it is necessary in some places to distinguish Number from Expr and so some limited support for that is included e.g.:
```python
from typing import reveal_type
from sympy import Symbol, Expr

def g(a: Expr, b: Expr):
    c1, _ = a.as_coeff_Mul()
    c2, _ = b.as_coeff_Mul()
    return c1 + c2

x = Symbol('x')
v = g(2*x, 3*x - 1)

reveal_type(v) # Number
```

#### Other comments

Apart from being able to recognise basic use of polys (gh-27218) the main thing needed after this is to make it so that evalf can be understood. There are still a huge number of errors under pyright and in fact I think that the specific change here actually increases the number of errors because it means that types are inferred in more places rather than being treated as Any/Unknown.

#### Release Notes

NO ENTRY

### CODE REVIEW

File: sympy/core/evalf.py [hunk 1]
```diff
@@ -1260,10 +1260,12 @@ def hypsum(expr: 'Expr', n: 'Symbol', start: int, prec: int) -> mpf:
...
         raise ValueError("Sum diverges like (n!)^%i" % (-h))
 
-    term = expr.subs(n, 0)
-    if not term.is_Rational:
+    eterm = expr.subs(n, 0)
+    if not eterm.is_Rational:
         raise NotImplementedError("Non rational term functionality is not implemented.")
 
+    term: Rational = eterm # type: ignore
+
```
Comment (sylee957) [inline sympy/core/evalf.py:1268]: I think that `is_Rational` is not useful for type guarding.
Maybe [PEP-647](https://peps.python.org/pep-0647/) may help, but I don't think that it is possible, or at least easy how to make `obj.is_Rational` work.

However, the suggestion is that could we just use `isinstance(term, Rational)` if the performance changes of using `isinstance` is asymptotically not significant for the case.

File: sympy/core/numbers.py [hunk 2]
```diff
@@ -440,8 +442,16 @@ def class_key(cls):
...
+        raise NotImplementedError
+
+    @overload
+    def __add__(self, other: Number | int | float) -> Number: ...
+    @overload
+    def __add__(self, other: Expr) -> Expr: ...
+
     @_sympifyit('other', NotImplemented)
-    def __add__(self, other):
+    def __add__(self, other) -> Expr:
```
[the remark this answers is not in the document]
Comment (oscarbenjamin) [reply sympy/core/numbers.py:454]: I don't see that error from pyright either in the editor or when run from command line. Maybe it is something to do with pylance.

There will be some cases where new errors are shown when adding type hints. The hints added here are not intended to suppress all errors from pyright in core but rather to improve inference elsewhere:
```python
from sympy import *
x = Symbol('x')
e = x + x
n, p = e.as_coeff_Mul()
reveal_type(n + n) # Number
```

File: sympy/core/evalf.py (hunk 1 above)
Comment (oscarbenjamin) [reply sympy/core/evalf.py:1268]: This example could use `isinstance` but there are many other cases like this throughout the codebase and I am not sure how best to handle them in general. I have only changed these when mypy reports an error. Currently mypy checks this function but infers that `subs` returns `Any`. The change in this PR means that it infers that `subs` returns `Expr` and then it complains that `Expr` does not have a `.p` attribute.

**Comment by sylee957** [comment]
I'm okay with the changes. But it looks like the problem is that `type: ignore` comments are added a lot, even for some existing annotations. I think that workaround code should be reduced in the future eventually.

**Comment by oscarbenjamin** [comment]
Maybe we should start an issue to discuss how the hints could theoretically work in a clean codebase.

**Pull request merged**